Fix deprecation notices Indexes annotation

jmikola opened the pull request:
<!-- Fill in the relevant information below to help triage your pull request. -->

|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Related issues | #2235, #2249

#### Summary

The `Indexes` attribute/annotation was deprecated in [doctrine/mongodb-odm@7435e66](https://github.com/doctrine/mongodb-odm/commit/7435e66b974130931b3a4acadb492fe572ff9298) for 2.2. There were deprecation messages added for using `Indexes` at the class level, and `indexes` within the Document annotation; however, it looks like Indexes on the property level was [missed](https://github.com/doctrine/mongodb-odm/blob/7435e66b974130931b3a4acadb492fe572ff9298/lib/Doctrine/ODM/MongoDB/Mapping/Driver/AnnotationDriver.php#L189).

Also, one of the deprecation messages got screwed up in [doctrine/mongodb-odm@3bdfe97](https://github.com/doctrine/mongodb-odm/commit/3bdfe97be472160ebbc924d5882ce61bd640a515#diff-65de5973d8c41abd6a3cc14c862908909d49367a01216544630b343b500929d1R153). The message changed but the sprintf args weren't swapped to account for that.

I struggled to add tests for this because:

 * `trigger_deprecation` calls `@trigger_error()`, so output is silenced. I could theoretically use [`error_get_last()`](https://www.php.net/manual/en/function.error-get-last.php) to check for the deprecation message, but there's no prior art for doing so. In fact, no deprecation messages seem to be tested.
 * The `Indexes` attribute cannot target a property, so the property-level deprecation message could only trigger when using it as annotation.
 * I couldn't discern the syntax for using `Indexes` as a class-level attribute with `Index` nested within. I consulted doctrine/orm#9241 some some examples of nesting attributes, but all of my attempts resulted in exceptions (e.g. `$value` property of `Indexes` couldn't be set).

For manual testing of the annotation form, this is what I used:

```php
/**
 * @ODM\Document
 * @ODM\Indexes({
 *   @ODM\Index(keys={"foo"="asc"})
 * })
 */
class DeprecatedIndexesAttributeOnClass
{
    /**
     * @ODM\Field(type="string")
     * @ODM\Indexes({
     *   @ODM\Index()
     * })
     */
    public string $foo;
}
```

jmikola committed:
Fix argument order for Indexes deprecation notice

The argument order has been incorrect since the deprecation message was changed in 3bdfe97be472160ebbc924d5882ce61bd640a515.

jmikola committed:
Deprecation notice when using Indexes annotation on a property

This was missed when Indexes was originally deprecated in 7435e66b974130931b3a4acadb492fe572ff9298.

IonBazan reviewed (approved):
I think we are missing tests for deprecation messages. Should we add them in this PR then?

jmikola commented:
> I think we are missing tests for deprecation messages. Should we add them in this PR then?

I'll make an attempt to do so before wrapping this up. We have some prior art for doing this in [PHPLIB](https://github.com/mongodb/mongo-php-library/blob/1.18.0/tests/TestCase.php#L188), but [Expecting Deprecations (E_USER_DEPRECATED)](https://docs.phpunit.de/en/11.1/error-handling.html#expecting-deprecations-e-user-deprecated) may have a better approach.

jmikola commented:
I made some progress in asserting deprecation messages, but ran into a roadblock attempting to use Indexes as an attribute (annotation is fine). I'll hold off on this until we can talk it through in the [Slack thread](https://doctrine.slack.com/archives/CAAE6T66B/p1713538516260789).

jmikola committed:
Test deprecation messages for Indexes annotation

The attribute is intentionally untested. The Indexes attribute cannot target a property (it omits Attribute::TARGET_PROPERTY). When Indexes targets a class, Index attributes cannot be nested within its definition. Testing the "indexes" option for class-level Document attributes is technically possible but not worth the effort.

jmikola commented:
Manually rebased on 2.7.x, since that's the same branch I used for previous fixes (https://github.com/doctrine/mongodb-odm/pull/2627).

jmikola commented:
Deprecation messages for annotations are now tested. CI failures are unrelated to this PR and due to https://github.com/doctrine/mongodb-odm/issues/2624.

IonBazan commented on tests/Doctrine/ODM/MongoDB/Tests/Mapping/AnnotationDriverTest.php:
```diff
@@ -16,4 +26,125 @@ protected static function loadDriver(): MappingDriver
+    }
+
+    /** @param list<string>|null $errors */
+    private function captureDeprecationMessages(callable $callable, ?array &$errors): mixed
```
I think we can also use https://symfony.com/doc/current/components/phpunit_bridge.html but seems like a drastic change so let's keep it this way for now.

jmikola replied:
Aye. We use that on the MongoDB driver but I noted it wasn't being utilized here.
